<a href="https://colab.research.google.com/github/Ricky0917/STAT-545/blob/main/worksheet_b01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Name:

Student ID:

# Worksheet B1: Writing functions & tests (8 points)
**Fall 2025**

There are 8 questions for Worksheet B01, each worth one point.

## Getting Started

Load the requirements for this worksheet:

In [13]:
suppressPackageStartupMessages(library(palmerpenguins))
suppressPackageStartupMessages(library(lubridate))
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(testthat))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(gapminder))
suppressPackageStartupMessages(library(digest))

The following code chunk has been unlocked, to give you the flexibility to start this document with some of your own code. Remember, it's bad manners to keep a call to `install.packages()` in your source code, so don't forget to delete these lines if you ever need to run them.

In [12]:
# An unlocked code cell.
install.packages("palmerpenguins")
install.packages("gapminder")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



## QUESTION 1

{points: 1}

Create a function that allows you to compute the max minus min of the Adelie penguins body mass. A code snippet to calculate max minus min *without a function* is shown to help you - essentially, your task is to turn this code snippet into a function. Be sure to make all three objects shown below.

```r
# put into practice your knowledge of dplyr to subset the penguins dataset to only those from the Adelie species
adelie <- penguins %>%
    filter(FILL_THIS_IN == FILL_THIS_IN) %>%
    drop_na()

# write your function here
max_minus_min <- function(x) FILL_THIS_IN - FILL_THIS_IN

# apply your function to the Adelie penguins body mass
answer1.0 <- max_minus_min(FILL_THIS_IN)
```

In [8]:
# your code here
# put into practice your knowledge of dplyr to subset the penguins dataset to only those from the Adelie species
adelie <- penguins %>%
  filter(species == "Adelie") %>%
  drop_na()

# write your function here
max_minus_min <- function(x) max(x) - min(x)

# apply your function to the Adelie penguins body mass
answer1.0 <- max_minus_min(adelie$body_mass_g)
#fail() # No Answer - remove if you provide an answer
head(adelie)
cat("Your final answer:")
answer1.0

species island    bill_length_mm bill_depth_mm flipper_length_mm body_mass_g
1 Adelie  Torgersen 39.1           18.7          181               3750       
2 Adelie  Torgersen 39.5           17.4          186               3800       
3 Adelie  Torgersen 40.3           18.0          195               3250       
4 Adelie  Torgersen 36.7           19.3          193               3450       
5 Adelie  Torgersen 39.3           20.6          190               3650       
6 Adelie  Torgersen 38.9           17.8          181               3625       
  sex    year
1 male   2007
2 female 2007
3 female 2007
4 female 2007
5 male   2007
6 female 2007

Your final answer:

[1] 1925

In [9]:
library(digest)
stopifnot("adelie should be a data frame"= setequal(digest(paste(toString('data.frame' %in% class(adelie)), "bb056")), "a39d76d6296a3037e78b6263f1d9f43e"))
stopifnot("dimensions of adelie are not correct"= setequal(digest(paste(toString(dim(adelie)), "bb056")), "a5f5671797f20e10825452e32ff073da"))
stopifnot("column names of adelie are not correct"= setequal(digest(paste(toString(sort(colnames(adelie))), "bb056")), "70b631c0e229f1668f78e1516adc8c03"))
stopifnot("types of columns in adelie are not correct"= setequal(digest(paste(toString(sort(unlist(sapply(adelie, class)))), "bb056")), "32b76af8c4c662422fe9fb58496f6f46"))
stopifnot("values in one or more numerical columns in adelie are not correct"= setequal(digest(paste(toString(if (any(sapply(adelie, is.numeric))) sort(round(sapply(adelie[, sapply(adelie, is.numeric)], sum, na.rm = TRUE), 2)) else 0), "bb056")), "b4b2079a5362cde6d86df445ff99d30e"))
stopifnot("values in one or more character columns in adelie are not correct"= setequal(digest(paste(toString(if (any(sapply(adelie, is.character))) sum(sapply(adelie[sapply(adelie, is.character)], function(x) length(unique(x)))) else 0), "bb056")), "ff650c9ac0ece25d68034ee4839b14e3"))
stopifnot("values in one or more factor columns in adelie are not correct"= setequal(digest(paste(toString(if (any(sapply(adelie, is.factor))) sum(sapply(adelie[, sapply(adelie, is.factor)], function(col) length(unique(col)))) else 0), "bb056")), "b1639959321a46bd484921a4d0bd2a22"))

stopifnot("type of answer1.0 is not integer"= setequal(digest(paste(toString(class(answer1.0)), "bb057")), "58d461f589d693fde7371bf55655762f"))
stopifnot("length of answer1.0 is not correct"= setequal(digest(paste(toString(length(answer1.0)), "bb057")), "10ecda7ae540547b1f228fc02a55c49b"))
stopifnot("values of answer1.0 are not correct"= setequal(digest(paste(toString(sort(answer1.0)), "bb057")), "caa27b2a9fec7bf0e5c830d8160d6cbc"))

print('Success!')

[1] "Success!"


## QUESTION 2

{points: 1}


Test your function on the life expectancy variable of the `gapminder` dataset and assign the returned value to R object `answer2.0`. Does it work?

In [10]:
# your code here
answer2.0 <- max_minus_min(gapminder$lifeExp)
#fail() # No Answer - remove if you provide an answer
print(answer2.0)

[1] 59.004


In [11]:
library(digest)
stopifnot("type of answer2.0 is not numeric"= setequal(digest(paste(toString(class(answer2.0)), "2f9f7")), "a089c1e682315054f6e6ffc338d129ea"))
stopifnot("value of answer2.0 is not correct (rounded to 2 decimal places)"= setequal(digest(paste(toString(round(answer2.0, 2)), "2f9f7")), "1600bf62fa177fbf3aa6651857df1a7d"))
stopifnot("length of answer2.0 is not correct"= setequal(digest(paste(toString(length(answer2.0)), "2f9f7")), "e8c6b0f50e51743c7099a877fbc4ece5"))
stopifnot("values of answer2.0 are not correct"= setequal(digest(paste(toString(sort(round(answer2.0, 2))), "2f9f7")), "1600bf62fa177fbf3aa6651857df1a7d"))

print('Success!')

[1] "Success!"


## QUESTION 3

{points: 1}


The function shouldn't work with all inputs, such as the following examples:

```
max_minus_min(adelie)
max_minus_min(penguins$species)
max_minus_min("stat545 is great")
```

We expect errors for these, because it wouldn't make sense to compute the max minus min in any of those arguments.

However, R will often try and make sense of your function... and that is not always a good thing. If you run the code below, you will see that the function is giving us an output for arguments that do not make any sense.

In [12]:
max_minus_min(gapminder[c('lifeExp', 'gdpPercap', 'pop')])
max_minus_min(c(TRUE, FALSE, TRUE))

[1] 1318683072

[1] 1

To avoid this from happening, rewrite the `max_minus_min` function to include a `stopifnot()`. Think of what should be the argument of `stopifnot()` to solve this issue - in other words, what should be the object class that the function `max_minus_min` should accept (e.g. a dataframe, a character, a number...)?

```r
answer3.0 <- function(x) {
  stopifnot(FILL_THIS_IN)
  max(x) - min(x)
}
```

In [13]:
# your code here
answer3.0 <- function(x) {
  stopifnot(is.numeric(x))
  max(x) - min(x)
}

#fail() # No Answer - remove if you provide an answer

In [14]:
library(digest)
stopifnot("type of (answer3.0(c(2,5,7)) == 5) is not logical"= setequal(digest(paste(toString(class((answer3.0(c(2,5,7)) == 5))), "a6be9")), "b70935854e84423ab528f01299627852"))
stopifnot("logical value of (answer3.0(c(2,5,7)) == 5) is not correct"= setequal(digest(paste(toString((answer3.0(c(2,5,7)) == 5)), "a6be9")), "6143ecb0dd777b3745092fb562c025f2"))

stopifnot("type of test_that(\"Question 3\", {expect_error(answer3.0(c('4', '20', '12')))}) == TRUE is not logical"= setequal(digest(paste(toString(class(test_that("Question 3", {expect_error(answer3.0(c('4', '20', '12')))}) == TRUE)), "a6bea")), "18f4f9c6157cd90435494c94211f887f"))
stopifnot("logical value of test_that(\"Question 3\", {expect_error(answer3.0(c('4', '20', '12')))}) == TRUE is not correct"= setequal(digest(paste(toString(test_that("Question 3", {expect_error(answer3.0(c('4', '20', '12')))}) == TRUE), "a6bea")), "b8f3e1c04815dbb4fd8b4393cc7442a0"))

stopifnot("type of test_that(\"Question 3\", {expect_error(answer3.0(list(1:5)))}) == TRUE is not logical"= setequal(digest(paste(toString(class(test_that("Question 3", {expect_error(answer3.0(list(1:5)))}) == TRUE)), "a6beb")), "107ebeb13b6291eb0a099cd68be42c92"))
stopifnot("logical value of test_that(\"Question 3\", {expect_error(answer3.0(list(1:5)))}) == TRUE is not correct"= setequal(digest(paste(toString(test_that("Question 3", {expect_error(answer3.0(list(1:5)))}) == TRUE), "a6beb")), "c12e6d768323331f16aec886b6e5b220"))

print('Success!')

Test passed 🥇
Test passed 😀
Test passed 🎉
Test passed 😀
[1] "Success!"


## QUESTION 4

{points: 1}


In the following chunk, I created a function to convert fahrenheit to celsius. When I test it with arguments that theoretically shouldn't give an output, it appears to have the same problem as our original `max_minus_min()` function.

In [14]:
# here I write the function
fahrenheit_to_celsius <- function(temp_F) {
  (temp_F - 32) * 5 / 9
}

In [15]:
# here I test it on something that theoretically should not work.. but does
fahrenheit_to_celsius(c(TRUE, FALSE, FALSE, TRUE))

[1] -17.22222 -17.77778 -17.77778 -17.22222

Rewrite the `fahrenheit_to_celsius` function using `if()` and `stop()` instead of `stopifnot()`. This allows you to write your own (more informative) error message.

*Hint:* Remember that you are trying to stop the function from working **if the argument is not numeric.**

```r
answer4.0 <- function(temp_F) {
  if(!FILL_THIS_IN(FILL_THIS_IN)) {
    stop('I am so sorry, but this function only works for numeric input!\n',
         'You have provided an object of class: ', class(FILL_THIS_IN)[1])
  }
  temp_C <- (temp_F - 32) * 5 / 9
  return(FILL_THIS_IN)
}
```

In [16]:
# your code here
answer4.0 <- function(temp_F) {
  if(!is.numeric(temp_F)) {
    stop('I am so sorry, but this function only works for numeric input!\n',
         'You have provided an object of class: ', class(temp_F)[1])
  }
  temp_C <- (temp_F - 32) * 5 / 9
  return(temp_C)
}

#fail() # No Answer - remove if you provide an answer

In [17]:
library(digest)
stopifnot("type of (answer4.0(90) == 32.22222) is not logical"= setequal(digest(paste(toString(class((answer4.0(90) == 32.22222))), "dc5fc")), "95d1d5ce954dff8b7b80651538fcb559"))
stopifnot("logical value of (answer4.0(90) == 32.22222) is not correct"= setequal(digest(paste(toString((answer4.0(90) == 32.22222)), "dc5fc")), "d90a7babb2a1decedb83deb7ebca1bb7"))

stopifnot("type of (answer4.0(c(100, 30)) == c(37.777778, -1.111111)) is not logical"= setequal(digest(paste(toString(class((answer4.0(c(100, 30)) == c(37.777778, -1.111111)))), "dc5fd")), "60ab00c1814bcda50630930f326243aa"))
stopifnot("logical value of (answer4.0(c(100, 30)) == c(37.777778, -1.111111)) is not correct"= setequal(digest(paste(toString((answer4.0(c(100, 30)) == c(37.777778, -1.111111))), "dc5fd")), "47b5e3f3404f346c9e0bfb1ed539f1af"))

stopifnot("type of (test_that(\"Question 4\", {expect_error(answer4.0(\"90\"))}) == TRUE) is not logical"= setequal(digest(paste(toString(class((test_that("Question 4", {expect_error(answer4.0("90"))}) == TRUE))), "dc5fe")), "09f415a5e9061005186d9a5a4c049ada"))
stopifnot("logical value of (test_that(\"Question 4\", {expect_error(answer4.0(\"90\"))}) == TRUE) is not correct"= setequal(digest(paste(toString((test_that("Question 4", {expect_error(answer4.0("90"))}) == TRUE)), "dc5fe")), "dada4aa4ef670a3737ef35fe31696429"))

stopifnot("type of (test_that(\"Question 4\", {expect_error(answer4.0(list(1:5)))}) == TRUE) is not logical"= setequal(digest(paste(toString(class((test_that("Question 4", {expect_error(answer4.0(list(1:5)))}) == TRUE))), "dc5ff")), "1756afc145c7bb3c76b32796d1e3d3e1"))
stopifnot("logical value of (test_that(\"Question 4\", {expect_error(answer4.0(list(1:5)))}) == TRUE) is not correct"= setequal(digest(paste(toString((test_that("Question 4", {expect_error(answer4.0(list(1:5)))}) == TRUE)), "dc5ff")), "d78c20e95761a19c2f58c031f501f504"))

print('Success!')

Test passed 😀
Test passed 😀
Test passed 🎊
Test passed 🥳
[1] "Success!"


## QUESTION 5

{points: 1}


Write a function that takes 2 arguments (name them `x` and `y`), raises the first argument to the power of the second one, and outputs a string with a message that indicates what the output of the function is. Use `stopifnot()` so that the function *only* takes numeric arguments.

```r
# function to write text output of x raised to the power y.
answer5.0 <- function(FILL_THIS_IN, FILL_THIS_IN) {
    stopifnot(FILL_THIS_IN && FILL_THIS_IN)
    result <- FILL_THIS_IN
    paste(FILL_THIS_IN, "raised to the power of", FILL_THIS_IN, "is", FILL_THIS_IN)
}
```

In [18]:
# your code here
answer5.0 <- function(x, y) {
  stopifnot(is.numeric(x) && is.numeric(y))
  result <- x ^ y
  paste(x, "raised to the power of", y, "is", result)
}
#fail() # No Answer - remove if you provide an answer
answer5.0(3, 4)

[1] "3 raised to the power of 4 is 81"

In [19]:
library(digest)
stopifnot("type of answer5.0(3, 4) == 81 is not logical"= setequal(digest(paste(toString(class(answer5.0(3, 4) == 81)), "b96d5")), "73c923260df1e3ae69d1760f8006a232"))
stopifnot("logical value of answer5.0(3, 4) == 81 is not correct"= setequal(digest(paste(toString(answer5.0(3, 4) == 81), "b96d5")), "307983eab47e6068e18e633c7e2aa010"))

stopifnot("type of test_that(\"Question 5\", {expect_error(answer5.0(40, \"90\"))}) == TRUE is not logical"= setequal(digest(paste(toString(class(test_that("Question 5", {expect_error(answer5.0(40, "90"))}) == TRUE)), "b96d6")), "dc97cfc323a6855fdd9ab8dc34db52d6"))
stopifnot("logical value of test_that(\"Question 5\", {expect_error(answer5.0(40, \"90\"))}) == TRUE is not correct"= setequal(digest(paste(toString(test_that("Question 5", {expect_error(answer5.0(40, "90"))}) == TRUE), "b96d6")), "44282d2f20b640a2be1c023fa53c27ef"))

print('Success!')

Test passed 🥇
Test passed 🎊
[1] "Success!"


## QUESTION 6

{points: 1}


Use `expect_equal()` to check that manually calculating the max minus min of the bill length of all penguins in the `penguins` dataset yields the same result as using a function.

Run the following chunk of code first.

In [20]:
x <- max(penguins$bill_length_mm, na.rm = TRUE)
y <- min(penguins$bill_length_mm, na.rm = TRUE)
x - y

[1] 27.5

I have written a slightly tweaked version of the `max_minus_min()` function created in **Q1** that allows the user to control the behaviours around NAs. This is important - without removing NAs, if you try running `max_minus_min()` with `penguins$bill_length_mm` as an argument, the output will be NA.

In [21]:
# new function
max_minus_min2 <- function(x, na.rm = TRUE) {
  if(!is.numeric(x)) {
    stop('I am so sorry, but this function only works for numeric input!\n',
         'You have provided an object of class: ', class(x)[1])
  }
  max(x, na.rm = na.rm) - min(x, na.rm = na.rm)
}

Now, check that the output of this function when calculating the max minus min of the bill length across all penguins in the `penguins` dataset is the same as calculating it "manually" as we did above - in other words, that the function output **equals** 27.5.

__NOTE__: We're _only_ getting you to store the output of `expect_equal()` in variable `answer6.0` so that we can run the autograder! Otherwise, you'd only ever run it on its own, without assigning it to anything.

_Psst... take a look at the test cells in these worksheets. Notice a similarity?_

```r
answer6.0 <- expect_equal(FILL_THIS_IN, FILL_THIS_IN(FILL_THIS_IN$FILL_THIS_IN, na.rm = TRUE))
```

In [24]:
# your code here
answer6.0 <- expect_equal(x - y, max_minus_min2(penguins$bill_length_mm, na.rm = TRUE))

#fail() # No Answer - remove if you provide an answer
answer6.0

[1] 27.5

In [25]:
library(digest)
stopifnot("type of answer6.0 is not numeric"= setequal(digest(paste(toString(class(answer6.0)), "888d9")), "bff4a85aff4a2e704b34abb36aa27e41"))
stopifnot("value of answer6.0 is not correct (rounded to 2 decimal places)"= setequal(digest(paste(toString(round(answer6.0, 2)), "888d9")), "f0bd19551a9e918fe2ce2bea16dc4e91"))
stopifnot("length of answer6.0 is not correct"= setequal(digest(paste(toString(length(answer6.0)), "888d9")), "0201edfc71ee6793de8a1a59462b1d53"))
stopifnot("values of answer6.0 are not correct"= setequal(digest(paste(toString(sort(round(answer6.0, 2))), "888d9")), "f0bd19551a9e918fe2ce2bea16dc4e91"))

print('Success!')

[1] "Success!"


There are other `expect_` functions such as `expect_identical()`, `expect_match()` or `expect_output()` that you can look into further [here](https://r-pkgs.org/tests.html#test-structure).

## QUESTION 7

{points: 1}


Let's try combining a different expectation function and `test_that()` with a very easy example. We start by writing a very simple function that returns the string "Hello world, my name is" + your name.

In [26]:
hello_world <- function(x) {
    stopifnot(is.character(x)) # we want x to be a character
    paste('Hello world, my name is', x)
}

In the next code cell, we'll run `hello_world()` with "Julie Payette" as an argument, just to see the output!

In [27]:
hello_world("Julie Payette")

[1] "Hello world, my name is Julie Payette"

Great! Now, let's simply test that the output of `hello_world()` with _your_ name as an argument matches the character vector **'Hello world, my name is'**. I have added an example of how `expect_match()` works to help you write your test.

```
eggplants <- "Eggplants are purple"
expect_match(eggplants, "Eggplants") # works
expect_match(eggplants, "purple") # works

# your turn
answer7.0 <- test_that("returns hello world + your name string", {
    expect_match(FILL_THIS_IN(FILL_THIS_IN), FILL_THIS_IN)
})
```

In [28]:
# your code here
answer7.0 <- test_that("returns hello world + your name string", {
  expect_match(hello_world("Ricky Zng"), "Hello world, my name is")
})

#fail() # No Answer - remove if you provide an answer

Test passed 🎊


In [29]:
library(digest)
stopifnot("type of answer7.0 is not logical"= setequal(digest(paste(toString(class(answer7.0)), "854ab")), "bd2ffb22ebe852fe1a541cac6e23b33d"))
stopifnot("logical value of answer7.0 is not correct"= setequal(digest(paste(toString(answer7.0), "854ab")), "1487489f0c1411790cdab002886898b1"))

print('Success!')

[1] "Success!"


## QUESTION 8

{points: 1}



Create a function called `m` that multiplies two numbers (arguments `x` and `y`). Create a test for function `m` with description "Testing multiplication function" and add a few scenarios to it:

+ Check if `m(2,3)` equals `6`
+ Check if `m(2, c(2,3))` equals `c(4,6)`
+ Check if `m(2, "3")` throws an error "non-numeric argument to binary operator"

```r
m <- function(FILL_THIS_IN) FILL_THIS_IN
answer8.0 <- test_that("Testing multiplication function", {
  expect_equal(FILL_THIS_IN)
  expect_equal(FILL_THIS_IN)
  expect_error(FILL_THIS_IN, FILL_THIS_IN)
})
```

In [30]:
# your code here
m <- function(x, y) x * y

answer8.0 <- test_that("Testing multiplication function", {
  expect_equal(m(2, 3), 6)
  expect_equal(m(2, c(2, 3)), c(4, 6))
  expect_error(m(2, "3"), "non-numeric argument to binary operator")
})

#fail() # No Answer - remove if you provide an answer

Test passed 🎉


In [31]:
library(digest)
stopifnot("type of answer8.0 is not logical"= setequal(digest(paste(toString(class(answer8.0)), "8a0c7")), "1b6d603d6d15196cc0d6f7c08867c66c"))
stopifnot("logical value of answer8.0 is not correct"= setequal(digest(paste(toString(answer8.0), "8a0c7")), "1e930cf6c0f2a9a8fe6473fe6a1318b8"))

print('Success!')

[1] "Success!"


That's all! 👏

Please restart your kernal and re-run your code from top to bottom to ensure that no errors persist. Save your worksheet and submit the .ipynb file to Canvas.